Add File Service

marl0vv opened the pull request:
Добавил File Service. Идея в том, что контроллер будет передавать полученные файлы для записи в сервис и получать их для обработки из сервиса. При вызове конструктора File Service создаст в Temp структуру папок вида:
- webconverter-18333313453830189902
    - in
    - out

Метод saveInputFile сохраняет исходные файлы в in с уникальным именем (с помощью UUID).

marl0vv committed:
Add file service

Add file service that manages creation of temp webpconverter- folder that contains "in" and "out" folders and saves input files.

marl0vv committed:
Take out save file logic from Controller

Take out logic of saving files to FileService

RavenVSS commented on src/main/java/com/glamik/webpconverter/controller/ConverterAsyncController.java:
```diff
@@ -13,29 +15,23 @@
 public class ConverterAsyncController {
 
     private final ConversionTaskService conversionTaskService;
+    private final FileService fileService;
```
Когда идёт взаимодействие с несколькими сервисами, желательно использовать паттерн Команда.  

Почитать можно тут:  
- https://www.baeldung.com/java-command-pattern
- https://springframework.guru/gang-of-four-design-patterns/command-pattern/

marl0vv replied:
Почитал, добавил. Не уверен, что корректно применил, но вроде выглядит так, как описано в статьях.

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/FileService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private File createBaseDirectory() throws IOException {
+        File tempDir = File.createTempFile(BASE_DIR, "");
```
ОС имеет свойство удалять по своему усмотрению временные файлы, для нас это не желательно, поэтому файлы лучше хранить в каком-нибудь надёжном месте  

Например рядом с сервисом

marl0vv replied:
Не знаю, почему, но я с чего-то решил, что ОС не чистит временные файлы :D
Поменял местоположение на то же место, где лежит папка с самой программой.

marl0vv committed:
Use Command pattern in Controller

Use Command pattern in Controller to save input file and write to database about conversion task

marl0vv committed:
Change location of a folder containing images

Take out folder with images out of temp because it can be deleted by Windows

RavenVSS commented on src/main/java/com/glamik/webpconverter/command/Command.java:
```diff
@@ -0,0 +1,5 @@
+package com.glamik.webpconverter.command;
+
+public interface Command {
+    void execute();
```
Можно немного прокачать интерфейс, сделать его с обобщёнными типами ([Java Generics](https://metanit.com/java/tutorial/3.11.php))

Т.е. в качестве аргумента будет тип Arg, а за место void - тип Return

RavenVSS commented on src/main/java/com/glamik/webpconverter/command/SaveConversionTaskCommand.java:
```diff
@@ -0,0 +1,21 @@
+
+@RequiredArgsConstructor
+@Getter
+public class SaveConversionTaskCommand implements Command {
```
Если выше добавить обобщения и тут добавить в качестве аргумента тип `MultipartFile`, а возвращаемое значение `ConversionTask`, можно будет сделать класс бином Spring через аннотацию `@Component`

Далее внедрить его в контроллер и вызывать метод `execute`.  

Только в твоём случае в одной Команде должно быть и сохранение файла и создание `ConversionTask`

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/FileService.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private File createBaseDirectory() throws IOException {
+        String currentDir = System.getProperty("user.dir");
```
Я в задаче оставлял ссылку на то, как получать конфиги через аннотацию `@Value`, попробуй через неё получать путь до базовой директории, т.е, чтобы `BASE_DIR` был настраиваемым, а не создавать

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/FileService.java:
```diff
@@ -0,0 +1,70 @@
+        return new File(baseDir, IN_DIR);
+    }
+
+    public File getOutDir() {
```
Методы, которые не должны смотреть во внешний мир, лучше сделать `private`

marl0vv committed:
Fix getting directory from code to config file

marl0vv committed:
Change getting folder paths method to private

marl0vv committed:
Change command to be generic

Change command to generics and combine SaveInputFile command and SaveConversionTask command classes in one

RavenVSS commented on src/main/java/com/glamik/webpconverter/service/FileService.java:
```diff
@@ -0,0 +1,74 @@
+
+    public File saveInputFile(MultipartFile imageFile) throws IOException {
+        String originalFilename = Objects.requireNonNull(imageFile.getOriginalFilename(), "File must have a name");
+        String fileExtension = getFileExtension(originalFilename);
```
Вынести в Command

RavenVSS commented on src/main/java/com/glamik/webpconverter/controller/ConverterAsyncController.java:
```diff
@@ -1,41 +1,35 @@
-                .body(conversionTask.getId());
-    }
+            SaveConversionTaskCommand saveConversionTaskCommand = new SaveConversionTaskCommand(conversionTaskService, fileService);
+            saveConversionTaskCommand.execute(imageFile);
```
Создать Command через `@Component` и сюда добавить как bean через конструктор

marl0vv committed:
Fix Command to be a Bean

Fix Command to be a Bean and injected into a Converter. Also delete some redundant variables and lines.

marl0vv committed:
Add new member to ConversionTask model

Add new member to model. Change ConversionTaskService to save this new member. Take out originalFileName to command in order to send it to both Services

marl0vv committed:
Fix minor things

Delete formatting problem
Delete Some warnings
Delete unused imports
